climada/engine/uncertainty/base.py

```diff
@@ -0,0 +1,728 @@
+        # zip_longest(flat_axes, cols, fillvalue=None):
+
+        nplots = len(self.distr_dict)
+        nrows, ncols = int(nplots / 3) + 1, min(nplots, 3)
+        fig, axis = plt.subplots(nrows=nrows, ncols=ncols, figsize=(20, 16))
+        if nplots > 1:
+            flat_axes = axis.flatten()
+        else:
+            flat_axes = [axis]
+        for ax, name_distr in zip_longest(flat_axes,
+                                    self.distr_dict.items(), 
+                                    fillvalue=None):
+            if name_distr is None:
+                ax.remove()
+                continue
+            (param_name, distr) = name_distr
+            x = np.linspace(distr.ppf(0.001), distr.ppf(0.999), 100)
+            ax.plot(x, distr.pdf(x), label=param_name)
+            ax.legend()
+        return fig, axis
+
+
+    def evaluate(self, uncvar_kwargs):
+        """
+        Evaluate the uncertainty variable.
+
+        Parameters
+        ----------
+        uncvar_kwargs : dict
+            These parameters will be passed to self.unc_var.
+            They must be the input parameters of the uncertainty variable .
+
+        Returns
+        -------
+
+            Evaluated uncertainty variable
+
+        """
+        return self.unc_var(**uncvar_kwargs)
+    
+    
+    @staticmethod
+    def var_to_uncvar(var):
+        """
+        Returns uncertainty variable with no distribution if var is not
+        an UncVar. Else, returns var.
+
+        Parameters
+        ----------
+        var : Object or else
+
+        Returns
+        -------
+        UncVar
+            var if var is UncVar, else UncVar with var and no distribution.
+
+        """
+
+        if isinstance(var, UncVar):
+            return var
+
+        return UncVar(unc_var=lambda: var, distr_dict={})
+
+
+
+class Uncertainty():
+    """
+    Uncertainty analysis class
+
+    This is the base class to perform uncertainty analysis on the outputs of a
+    climada.engine.impact.Impact() or climada.engine.costbenefit.CostBenefit()
+    object.
+
+    """
+
+    def __init__(self, unc_vars, sample=None, metrics=None,
+                 sensitivity=None):
+        """
+        Initialize Uncertainty
+
+        Parameters
+        ----------
+        unc_vars : dict
+            keys are names and values are climade.engine.uncertainty.UncVar 
+        sample : pd.DataFrame, optional
+            DataFrame of sampled parameter values. Column names must be
+            parameter names (all labels) from all unc_vars.
+            The default is pd.DataFrame().
+        metrics : dict(), optional
+            Dictionnary of the CLIMADA metrics for which the sensitivity
+            will be computed. For each sample, each metric must have a
+            definite value.
+            Keys are metric names (e.g. 'aai_agg', 'freq_curve') and 
+            values are pd.DataFrame with values for each parameter sample
+            (one row per sample).
+            The default is {}.
```

Comment:
It is a bad idea to have default iterable values and against the CLIMADA conventions, c.f. https://github.com/CLIMADA-project/climada_python/blob/main/doc/guide/Guide_PythonDos-n-Donts.ipynb

I found it clearer to say in the docstring what the actual default value is rather than None.

Reply:
Hmmm... maybe this was not translated from the user guides prior to the coding convention update. It should be added then.

Reply:
Actually, just found the reference is here:

https://github.com/CLIMADA-project/climada_python/blob/guides_update/doc/guide/Guide_Reviewer_Checklist.ipynb

Reply:
Ok!